pypif_sdk/func/calculate_funcs.py

```diff
@@ -0,0 +1,280 @@
+    """
+    n = 0
+    for a in elemental_array:
+        n += a["occurances"]
+    return n
+
+def _add_ideal_atomic_weights_(elemental_array):
+    """
+    Uses mendeleev.element to find the molar mass of the element in question, and then multiplies that by the occurances of the element.
+    Adds the "weight" property to each of the dictionaries in elemental_array
+    :param elemental_array an array of dictionaries containing information about the elements in the system 
+    Returns the appended elemental_array
+    """
+    for a in elemental_array:
+        # this_atomic_weight = element(a["symbol"]).atomic_weight
+        this_atomic_weight = elements_data[a["symbol"]]["atomic_weight"]
+        a["weight"] = round(a["occurances"] * this_atomic_weight, 2)
+    return elemental_array
+
+def _calculate_total_mass_(elemental_array):
+    """
+    Sums the total weight from all the element dictionaries in the elemental_array after each weight has been added (after _add_ideal_atomic_weights_())
+    :param elemental_array an array of dictionaries containing information about the elements in the system 
+    """
+    n = 0
+    for a in elemental_array:
+        n += a["weight"]
+    return n
+
+def _add_ideal_weight_percent_(elemental_array):
+    """
+    Adds the "weight_percent" property to each of the dictionaries in elemental_array
+    :param elemental_array an array of dictionaries containing information about the elements in the system 
+    Returns the appended elemental_array
+    """
+    t_mass = _calculate_total_mass_(elemental_array)
+    for a in elemental_array:
+        a["weight_percent"] = round(a["weight"]/t_mass * 100, 2)
+    return elemental_array
+
+def _create_emprical_compositional_array_(expanded_chemical_formaula_string):
+    """
+    Turns a chemical formula into an array of objects with element names and their corresponding number of occurances
+    :param expanded_chemical_formaula_string chemical formula string that is expanded and has no special characters
+    Returns an elemental_array (array of dictionaries) containing information on each element in the system and it's total number of occurances (no duplicates)
+    """
+    elemental_array = _create_compositional_array_(expanded_chemical_formaula_string)
+    condensed_array = _consolidate_elemental_array_(elemental_array)
+    return condensed_array
+
+def _add_atomic_percents_(elemental_array):
+    """
+    Adds ideal atomic percents to a emperical compositional element array generated using _create_emprical_compositional_array_()
+    :param elemental_array an array of dictionaries containing information about the elements in the system 
+    Returns the elemental_array with the atomic percent of each element added
+    """
+    n_atoms = _calculate_n_atoms_(elemental_array)
+    for e in elemental_array:
+        e["atomic_percent"] = round(e["occurances"]/n_atoms*100,2)
```

Comment:
Why round?  That makes the atomic percent composition non-equivalent to the original chemical formula.  The atomic percents are real-valued, anyways.